ext/strings.go

```diff
@@ -0,0 +1,628 @@
+			},
+			&functions.Overload{
+				Operator: "substring",
+				Binary:   callInStrIntOutStr(substr),
+				Function: callInStrIntIntOutStr(substrRange),
+			},
+			&functions.Overload{
+				Operator: "string_substring_int",
+				Binary:   callInStrIntOutStr(substr),
+			},
+			&functions.Overload{
+				Operator: "string_substring_int_int",
+				Function: callInStrIntIntOutStr(substrRange),
+			},
+			&functions.Overload{
+				Operator: "trim",
+				Unary:    callInStrOutStr(strings.TrimSpace),
+			},
+			&functions.Overload{
+				Operator: "string_trim",
+				Unary:    callInStrOutStr(strings.TrimSpace),
+			},
+			&functions.Overload{
+				Operator: "upper",
+				Unary:    callInStrOutStr(strings.ToUpper),
+			},
+			&functions.Overload{
+				Operator: "string_upper",
+				Unary:    callInStrOutStr(strings.ToUpper),
+			},
+		),
+	}
+}
+
+func after(str, sub string) (string, error) {
+	ind := strings.Index(str, sub)
+	if ind < 0 {
+		return "", nil
+	}
+	return str[ind+len(sub):], nil
+}
+
+func before(str, sub string) (string, error) {
+	ind := strings.Index(str, sub)
+	if ind < 0 {
+		return str, nil
+	}
+	return str[:ind], nil
+}
+
+func charAt(str string, ind int64) (string, error) {
+	i := int(ind)
+	if i < 0 || i >= len(str) {
+		return "", fmt.Errorf("index out of range: %d", ind)
+	}
+	return str[i : i+1], nil
+}
+
+func indexOf(str, substr string) (int64, error) {
+	return int64(strings.Index(str, substr)), nil
+}
+
+func indexOfOffset(str, substr string, offset int64) (int64, error) {
+	off := int(offset)
+	if off < 0 || off >= len(str) {
+		return -1, fmt.Errorf("index out of range: %d", off)
+	}
+	return offset + int64(strings.Index(str[offset:], substr)), nil
+}
+
+func replace(str, old, new string) (string, error) {
+	return strings.ReplaceAll(str, old, new), nil
+}
+
+func replaceN(str, old, new string, n int64) (string, error) {
+	return strings.Replace(str, old, new, int(n)), nil
+}
+
+func split(str, sep string) ([]string, error) {
+	return strings.Split(str, sep), nil
+}
+
+func splitN(str, sep string, n int64) ([]string, error) {
+	return strings.SplitN(str, sep, int(n)), nil
+}
+
+func substr(str string, start int64) (string, error) {
+	if int(start) < 0 || int(start) >= len(str) {
+		return "", fmt.Errorf("index out of range: %d", start)
+	}
+	return str[int(start):], nil
+}
+
+func substrRange(str string, start, end int64) (string, error) {
+	l := len(str)
+	if start > end {
+		return "", fmt.Errorf("invalid substring range. start: %d, end: %d", start, end)
+	}
+	if int(start) < 0 || int(start) >= l {
+		return "", fmt.Errorf("index out of range: %d", start)
+	}
+	if int(end) < 0 || int(end) >= l {
+		return "", fmt.Errorf("index out of range: %d", end)
+	}
+	return str[int(start):int(end)], nil
+}
+
+func callInStrOutStr(fn func(string) string) functions.UnaryOp {
+	return func(val ref.Val) ref.Val {
+		vVal, ok := val.(types.String)
+		if !ok {
+			return types.ValOrErr(val, "no such overload")
```

Comment:
One of these days a `"no such overlaod"` will sneak through.  Consider a constant for at least the string, and perhaps specialized versions of `NewErr()` or `ValOrErr()`.